core/client/fs/src/main/java/alluxio/client/file/UfsFileInStream.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.file;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Supports reading from a ufs file directly.
+ */
+// TODO(lu) make it thread safe
+@NotThreadSafe
+public class UfsFileInStream extends FileInStream {
```

Comment:
Is it possible to have multiple UFSes with different implementation? 
Anything common in the UFS layer?

Reply:
It's always possible, when one UFS implementation is mature, we can move to multiple. This also introduce a problem of how to support multiple UFSes
e.g. 
/folder
  /UFS1
  /UFS2
or a filesystem tree similar to master

that's something can be considered in the future

Reply:
and also if user wants multiple UFSes, they can launch alluxio fuse to mount different UFS to different local mount point, do not see the must-have for this feature